### PR TITLE
ENH: scipy.signal - Addition of CSD, coherence; Enhancement of welch

### PR DESCRIPTION
**Pull request by e-q**
This is a new PR for adding Cross Spectral Density computations to scipy.signal. The previous PR (#4468), extended the existing welch algorithm for the CSD, but then I found that matplotlib.malb has an existing, much faster, version that uses strided arrays for efficiency. 

I have adapted the matplotlib.mlab code for use in scipy.signal. This required rewriting welch to use the same underlying stride algorithm. 

Some reworking of the matplotlib code was needed to make functions work for
multidimensional arrays, as the current scipy.signal.welch function allows.
Things were also adjusted to keep the same welch API.

Advantages of this code change include:
- A speedup of about 8x in welch
- A speedup of about 6x in csd compared to extending the current
  algorithm to cover the cross spectral density.
- Straightforward extensibility to add spectrogram functionality

There are probably still rough edges, but this seems to be working. Some
tests are included. (Known test failure : test_complex_32 in
test_spectral.TestWelch; effectively correct, but the tolerances are too
tight.)

### CODE REVIEW

File: scipy/signal/spectral.py [hunk 1]
```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    elif mode == 'angle' or mode == 'phase':
+        result = np.angle(result)
+    elif mode == 'complex':
+        pass
+
+    result *= scale
+    if mode == 'psd' and sides == 'onesided':
+        result[...,1:-1] *= 2
+
+    t = np.arange(nfft/2, len(x) - nfft/2 + 1, nfft - noverlap)/fs
```
Comment (rgommers) [inline scipy/signal/spectral.py]: ``nfft/2`` here (2x) should be integer division it looks like

File: scipy/signal/spectral.py [hunk 2]
```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+
+    if not same_data:
+        # All the same operations on the y data
+        resultY = _stride_windows(y, nfft, noverlap, axis=-1)
+        resultY = detrend_func(resultY)
+        resultY = _apply_window(resultY, win, axis=-1)
+        padShape = list(resultY.shape)
+        padShape[-1] = nfft - padShape[-1]
+        resultY = np.concatenate((resultY, np.zeros(padShape)), axis=-1)
+        resultY = fftpack.fft(resultY, n=nfft, axis=-1)[...,:numFreqs]
```
Comment (rgommers) [inline scipy/signal/spectral.py]: You can refactor this into a helper function that returns ``result, pad_shape`` so you don't have to duplicate the above 8 lines

File: scipy/signal/spectral.py [hunk 3]
```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    result *= scale
+    if mode == 'psd' and sides == 'onesided':
+        result[...,1:-1] *= 2
+
+    t = np.arange(nfft/2, len(x) - nfft/2 + 1, nfft - noverlap)/fs
+
+    if sides == 'twosided':
+        pass
+        #freqs = fftpack.fftshift(freqs)
+        #result = fftpack.fftshift(result, axes=-1)
```
Comment (rgommers) [inline scipy/signal/spectral.py]: commented-out code should be removed or fixed.

**Comment by rgommers** [comment]
The private functions ``_stride_repeat``, ``_stride_windows`` and ``_apply_window`` are used exactly once, contain unnecessary checks (input validation should be done at the beginning of the public functions) and large docstrings that make the code hard to read. The actual functionality is only a couple of lines. I suggest to either remove these functions and simply add a comment inline about what ``as_strided`` is doing in each case, or at least strip them down to the minimum needed.

File: scipy/signal/spectral.py [hunk 4]
```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    window : str or tuple or array_like, optional
+        Desired window to use. See `get_window` for a list of windows and
+        required parameters. If `window` is array_like it will be used
+        directly as the window and its length will be used for nperseg.
+        Defaults to 'hanning'.
+    nperseg : int, optional
+        Length of each segment.  Defaults to 256.
+    noverlap: int, optional
+        Number of points to overlap between segments. If None,
+        ``noverlap = nperseg / 2``.  Defaults to None.
```
Comment (rgommers) [inline scipy/signal/spectral.py]: integer division?

Comment (rgommers) [inline scipy/signal/spectral.py]: this happens in a few more places, can you check the whole file?

File: scipy/signal/spectral.py [hunk 5]
```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    and on your requirements.  For the default 'hanning' window an
+    overlap of 50\% is a reasonable trade off between accurately estimating
+    the signal power, while not over counting any of the data.  Narrower
+    windows may require a larger overlap.
+    """
+    [ff,Pxx] = welch(x, fs, window, nperseg, noverlap, nfft, detrend, axis=axis)
+    [_, Pyy] = welch(y, fs, window, nperseg, noverlap, nfft, detrend, axis=axis)
+    [_, Pxy] = csd(x, y, fs, window, nperseg, noverlap, nfft, detrend, axis=axis)
+
+    Cxy = np.abs(Pxy)**2/Pxx/Pyy
```
Comment (rgommers) [inline scipy/signal/spectral.py]: PEP* nitpick: spaces around ``/``

File: scipy/signal/spectral.py (hunk 2 above)
Comment (rgommers) [inline scipy/signal/spectral.py]: and can you use PEP8 compliant names? So ``pad_shape``, ``result_y``, ``num_freqs``, etc.

File: scipy/signal/tests/test_spectral.py [hunk 6]
```diff
@@ -430,11 +430,291 @@ def test_complex_32(self):
...
+            0.11111111, 0.11111111, 0.11111111, 0.11111111, 0.07638889]))
+
+    def test_complex(self):
+        x = np.zeros(16, np.complex128)
+        x[0] = 1.0 + 2.0j
+        x[8] = 1.0 + 2.0j
+        f, p = csd(x, x, nperseg=8)
+        assert_allclose(f, fftpack.fftfreq(8, 1.0))
+        assert_allclose(p, np.array([0.41666667, 0.38194444, 0.55555556,
+            0.55555556, 0.55555556, 0.55555556, 0.55555556, 0.38194444]))
```
Comment (rgommers) [inline scipy/signal/tests/test_spectral.py]: style nitpick: can you align this like so

    np.array([0.41666667, ...
              0.55555556, ...

Comment (rgommers) [inline scipy/signal/tests/test_spectral.py]: and similarly, in several place:
    
    f, p = csd(x, ...
               detrend=...

**Comment by rgommers** [comment]
@e-q nice, timings look very promising. It would be good to add a benchmark for each function that is sped up here. This should be quite straightforward, see https://github.com/scipy/scipy/pull/4621 for an example

**Comment by e-q** [comment]
Thanks for the feedback! I've cleaned things up a bit; I suppose I wasn't critical enough of the original code's structure. 

Benchmarking sounds like a great idea. Oddly enough, these functions are now *slightly* slower than their matplotlib ancestors... I'll get to the bottom of it.

**Comment by rgommers** [comment]
Thanks, changes all look good - now this PR is pretty clean and easy (or easier) to review.

File: scipy/signal/spectral.py [hunk 7]
```diff
@@ -267,121 +267,495 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    # Detrend each data segment individually
+    result = detrend_func(result)
+    
+    # Create strided array of window profiles, same shape as result
+    if result.shape[-1] == 1:
+        winrep = win
+    else:
+        win_strides = np.zeros(len(result.shape), dtype='int64')
+        win_strides[-1] = win.strides[0]
+        winrep = np.lib.stride_tricks.as_strided(win, shape=result.shape,
```
Comment (ewmoore) [inline scipy/signal/spectral.py]: Is there a particular reason you've chosen to make a strided array of window profiles?  Since `win` is 1D and its length matches the last dimension of `result` you should be able to just use it directly. The multiplication will broadcast `win` across `result` in the usual manner.

File: scipy/signal/spectral.py [hunk 8]
```diff
@@ -267,121 +267,495 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    # Apply window by multiplication
+    result = winrep * result
+
+    # Zero pad here, now that windowing is done
+    pad_shape = list(result.shape)
+    pad_shape[-1] = nfft - pad_shape[-1]
+    result = np.concatenate((result, np.zeros(pad_shape)), axis=-1)
+
+    # Perform the fft
+    result = fftpack.fft(result, n=nfft, axis=-1)
```
Comment (ewmoore) [inline scipy/signal/spectral.py]: `fftpack.fft` will do the zero padding for you and defaults to operating along the last axis.  All the code to do zero padding can go away just above, and this could (but doesn't have to) become `result = fftpack.fft(result, n=nfft)`.

File: scipy/signal/spectral.py [hunk 9]
```diff
@@ -267,121 +267,495 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+    if x.ndim > 1:
+        if axis != -1:
+            x = np.rollaxis(x, axis, len(x.shape))
+            if not same_data and y.ndim > 1:
+                y = np.rollaxis(y, axis, len(y.shape))
+            # Output is going to have new last axis for window index
+            if axis < 0:
+                axis = len(np.broadcast(x,y).shape)-axis
 
     if x.size == 0:
```
Comment (ewmoore) [inline scipy/signal/spectral.py]: This check should come earlier.  Consider:
```
In [14]: a = np.array([[[],[],[]],[[],[],[]]])

In [15]: a.shape
Out[15]: (2, 3, 0)

In [16]: signal.welch(a, axis=1)
Out[16]:
(array([], shape=(2, 0, 3), dtype=float64),
 array([], shape=(2, 0, 3), dtype=float64))
```
Bit of corner case, I suppose.  Maybe this should be a test?  The currently shipping version of `welch` gets this right.

**Comment by e-q** [comment]
Good catches @ewmoore! Cleaning up that unnecessary work won an almost 10% speed boost. 

Indeed, the empty input cases are probably not so common, but I wouldn't want to cause any regression. To this end, I shuffled things around a bit to return the correctly broadcast shape for the CSD when only one input is empty, but can otherwise be broadcast (and tests for that case). Hopefully my implementation isn't too clunky...

I.e. ``csd(np.empty((10,10,3)), np.zeros((10,0,1)), axis=1)`` returns a shape of ``(10,0,3)``

**Comment by e-q** [comment]
I've written some simple benchmarks for `welch` and `csd`, that just run the functions on random arrays about a million items long. Comparing the new `welch` to the master branch on my machine:
```
BENCHMARK                                    BEFORE      AFTER     FACTOR
signal.CalculateWindowedFFT.time_welch       785.73ms    66.92ms   0.085169206x
```

The latest commit failed because `np.squeeze` doesn't take the `axis` kwarg in 1.6.2...

**Comment by ev-br** [comment]
Is https://github.com/matplotlib/matplotlib/issues/4324 anyhow relevant? (I've no idea; have just noticed that MPL issue, and this PR seems to use some code from roughly the same part of MPL...)

File: scipy/signal/spectral.py [hunk 10]
```diff
@@ -267,121 +267,498 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
...
+        result = np.conjugate(result) * result
+    elif mode == 'magnitude':
+        result = np.absolute(result)
+    elif mode == 'angle' or mode == 'phase':
+        result = np.angle(result)
+    elif mode == 'complex':
+        pass
+
+    result *= scale
+    if mode == 'psd' and sides == 'onesided':
```
Comment (tacaswell) [inline scipy/signal/spectral.py]: This line will need the same logic as was added in https://github.com/matplotlib/matplotlib/pull/4326

**Comment by tacaswell** [comment]
@ev-br Good catch!

**Comment by e-q** [comment]
The welch function in scipy 0.15 has the same bug that matplotlib/matplotlib#4324 describes, as can be found by running the same example code shown there, just with the appropriate call structure: 

`f, P = sig.welch(u, nperseg=u.size, fs=1/dt, window=sig.boxcar(u.size), detrend=None, noverlap=0, scaling='density', return_onesided=True)`

This means that the existing unit tests for welch, and the tests for CSD that I derived from them, that involve odd numbered fft segments will have to be changed as well.

**Comment by ev-br** [comment]
@tacaswell you did https://github.com/matplotlib/matplotlib/pull/4326, could you have a look on https://github.com/e-q/scipy/commit/853a4e20ad67472361fca3b177d7a7d7b959a24d please?

Just to be clear: I don't have an opinion about this PR (apart from noticing that the speedup looks nice), and I'll defer to @rgommers and @ewmoore for a review / decision on it.

**Comment by e-q** [comment]
Any suggestions what should still be worked on here? I'm eager to hear any additional feedback!

**Comment by rgommers** [comment]
@e-q a basic example with a plot in an ``Examples`` section in the docstrings of the new functions would be a nice bonus.

I'll play with this PR a little now; it seems to be in pretty good shape.

**Comment by rgommers** [comment]
@ewmoore do you still want/need to review the algorithm changes in more detail, or is this good to go from your POV?

**Comment by ewmoore** [comment]
Its probably fine. I'll have some time this weekend to look if you could
give another day or two.

**Comment by ewmoore** [comment]
The only thing I see is that I don't think the comment and reference to Bartlett's method really belongs anywhere but in `welch`.

**Comment by e-q** [comment]
Rebased, to take #4766 into account.

**Comment by ewmoore** [comment]
Nevermind, I must have been looking at an old version.  This looks good to me.

**Comment by rgommers** [comment]
LGTM too. Checked the benchmark, runs and for ``welch`` is indeed 8x faster than on current master. So in it goes. Thanks @e-q !

**Comment by e-q** [comment]
Ack, should I have squashed the commits first?

**Comment by rgommers** [comment]
It's fine like this. It's a pretty large PR that went through some changes; 17 commits is large-ish but not completely ridiculous.

**Comment by WarrenWeckesser** [comment]
@e-q Thanks for this PR, it is a nice addition to scipy.

One quibble (and sorry to get here late with this):  In the description of the `fs` argument, it should not say that the units are Hz.  Nothing in the code requires the units to be Hz, so saying that is confusing.  (E.g. if I have sunspot data that was recorded monthly, do I really have to convert my sample rate to Hz?)
